pandas/core/indexes/period.py

```diff
@@ -168,7 +168,7 @@ class PeriodIndex(DatetimeIndexOpsMixin, Int64Index, PeriodDelegateMixin):
     _is_numeric_dtype = False
     _infer_as_myclass = True
 
-    _data = None  # type: PeriodArray
+    _data = None
```

Comment:
So what is this getting inferred as now?

Reply:
@WillAyd  Inferred type is `None`, which is not correct, but putting the actual type is almost impossible. Because the type of `_data` in parent/sibling classes (e.g. in class `Index`, `DatetimeIndexOpsMixin`, etc.) is different. If we want to properly type this we need to create a type `ArrayLike`(I guess) in `pandas._typing` which can contain all possible type for `_data` then use it everywhere. But I don't know how many of these arraylike types(`PeriodArray`, `DatetimeLikeArray`, etc) are there, so someone will have to tell me how to find that, maybe @gfyoung can help with that.
If you want something like this, I would suggest doing that in a separate issue & PR as a lot of files will be changed as `_data` is in like 100 different files.

Reply:
We already have that in the typing module - did that not work?

Reply:
Okay, we do have array like in `_typing`, my bad, didn't saw that.
I think `_typing.AnyArrayLike` is what we need. Mypy will be happy with whatever type I put for `_data`, so someone should confirm that `AnyArrayLike` is the proper type to use for `_data`, and I will type `_data` in every class.

Reply:
Yea I think that should work here